runtime/security.go

```diff
@@ -609,6 +619,372 @@ func (p *securityEngine) applySecurityRuleRowFilter(res *ResolvedSecurity, _ *ru
+// Restricts access to these fields and rows by adding corresponding field access and row filter rules to the resolved security rules.
+func (p *securityEngine) resolveTransitiveAccessRuleForReport(ctx context.Context, instanceID string, claims *SecurityClaims, res *runtimev1.Resource) ([]*runtimev1.SecurityRule, error) {
+	var rules []*runtimev1.SecurityRule
+
+	report := res.GetReport()
+	if report == nil {
+		return nil, fmt.Errorf("resource is not a report")
+	}
+
+	spec := report.GetSpec()
+	if spec == nil {
+		return nil, fmt.Errorf("report spec is nil")
+	}
+	// explicitly allow access to the report itself
+	rules = append(rules, &runtimev1.SecurityRule{
+		Rule: &runtimev1.SecurityRule_Access{
+			Access: &runtimev1.SecurityRuleAccess{
+				Condition: fmt.Sprintf("'{{.self.kind}}'='%s' AND '{{lower .self.name}}'=%s", ResourceKindReport, duckdbsql.EscapeStringValue(strings.ToLower(res.Meta.Name.Name))),
+				Allow:     true,
+			},
+		},
+	})
+
+	// deny everything except the report itself, themes, explore, canvas and metrics view
+	var denyCondition strings.Builder
+	// self report
+	denyCondition.WriteString(fmt.Sprintf("('{{.self.kind}}'='%s' AND '{{lower .self.name}}'=%s)", ResourceKindReport, duckdbsql.EscapeStringValue(strings.ToLower(res.Meta.Name.Name))))
+	// all themes
+	denyCondition.WriteString(fmt.Sprintf(" OR '{{.self.kind}}'='%s'", ResourceKindTheme))
+
+	if spec.QueryName != "" {
+		initializer, ok := ResolverInitializers["legacy_metrics"]
+		if !ok {
+			return nil, fmt.Errorf("no resolver found for name 'legacy_metrics'")
+		}
+		resolver, err := initializer(ctx, &ResolverOptions{
+			Runtime:    p.rt,
+			InstanceID: instanceID,
+			Properties: map[string]any{
+				"query_name":      spec.QueryName,
+				"query_args_json": spec.QueryArgsJson,
+			},
+			Claims:    claims,
+			ForExport: false,
+		})
+		if err != nil {
+			return nil, err
+		}
+		defer resolver.Close()
+		rules = append(rules, resolver.InferRequiredSecurityRules()...)
+
+		// add refs to deny condition, relying on resolveRules to add the corresponding allow rules
+		mvName := ""
+		refs := resolver.Refs()
+		for _, ref := range refs {
+			// allow access to the referenced resource
+			rules = append(rules, &runtimev1.SecurityRule{
+				Rule: &runtimev1.SecurityRule_Access{
+					Access: &runtimev1.SecurityRuleAccess{
+						Condition: fmt.Sprintf("'{{.self.kind}}'=%s AND '{{lower .self.name}}'=%s", duckdbsql.EscapeStringValue(ref.Kind), duckdbsql.EscapeStringValue(strings.ToLower(ref.Name))),
+						Allow:     true,
+					},
+				},
+			})
+			// add to deny condition
+			denyCondition.WriteString(fmt.Sprintf(" OR ('{{.self.kind}}'=%s AND '{{lower .self.name}}'=%s)", duckdbsql.EscapeStringValue(ref.Kind), duckdbsql.EscapeStringValue(strings.ToLower(ref.Name))))
+			if ref.Kind == ResourceKindMetricsView {
+				mvName = ref.Name
+			}
+		}
```

Comment:
Can you think of any ways we could simplify some of this? Maybe with new security rules for resource access and/or for fallback access?

Reply:
You mean a way to not have a deny condition in addition to allow rule? That would be tricky because we add allow all access rule in `resolveRules` mainly to allow access to resource like mv, explore during usual access when there are not security rules defined inline in the resource. May be I can change allow all to just allow that resource that way we may not need the deny condition.